### PR TITLE
feat: support context cancellation under auto pipeline

### PR DESCRIPTION
**Pull request by rueian**
Support context deadline or cancellation with minimum overhead.

1. set context deadline to tcp connection while the `pipe` is in sync mode
2. check context channel if necessary while the `pipe` is in pipeline mode

### CODE REVIEW

File: pipe.go [hunk 1]
```diff
@@ -397,7 +450,12 @@ func (p *pipe) syncDo(cmd cmds.Completed) (resp RedisResult) {
 	return newResult(msg, err)
 }
 
-func (p *pipe) syncDoMulti(resp []RedisResult, multi []cmds.Completed) []RedisResult {
+func (p *pipe) syncDoMulti(ctx context.Context, resp []RedisResult, multi []cmds.Completed) []RedisResult {
+	if dl, ok := ctx.Deadline(); ok {
+		p.conn.SetDeadline(dl)
```
Comment (FZambia) [inline pipe.go:455]: This won't be called from different goroutines right? I.e. situation when multiple requests can concurrently modify connection deadline is impossible?

Comment (rueian) [reply pipe.go:455]: No it won't. 

Concurrently modify connection deadline is impossible because only one operation can enter the `syncDo` or `syncDoMulti` concurrently.

Comment (FZambia) [reply pipe.go:455]: Got it - code is a bit hard to follow - various states and atomics, let me ask one more question - there is a path which seems to queue requests (`p.queue.PutMulti(multi)`). I don't see how deadline on the connection is set for that path.

Comment (rueian) [reply pipe.go:455]: Deadline is not set on that path. Instead, it directly check if the context is canceled while waiting for responses.

https://github.com/rueian/rueidis/blob/4704eecaf90ec66d61afa0c5290280c3bcf66448/pipe.go#L408-L414

Comment (FZambia) [reply pipe.go:455]: I am worrying that deadline is not used there, let's say one request stuck in reading for some reason (ex. network), how this affect the entire pipeline?

Comment (rueian) [reply pipe.go:455]: Then the entire pipeline is also stuck.
If that is caused by network issue, the tcp keepalive should throw an error on the reading path and we should close the connection.

Comment (FZambia) [reply pipe.go:455]: I think the problem is that TCP keepalive may detect broken connection within pretty large time (up to 2 hours if I remember right).

Comment (rueian) [reply pipe.go:455]: https://github.com/rueian/rueidis/blob/d003ef93e0973a409b30b71f5b7df029ad13d119/rueidis.go#L23

I set the default interval to be 1 second. and the default of `tcp_keepalive_probes` in linux is 9. So, by default, it take 10 seconds to kill a connection. Maybe I could leave more documentation about the keepalive behavior.

Comment (rueian) [reply pipe.go:455]: Hi @FZambia,

Thank you for reading my code. The code is quite complex because I use two atomic variables `waits` and `state` to provide both sync mode (direct write&read on tcp connection) and auto pipeline mode (commands will be queue in buffer) and the two operation `Do` and `DoMulti` are also handled differently. I have tried to refactor them but not succeed yet.

Setting deadline on tcp connection under pipeline mode is hard. I still couldn't find a way to deal with deadline error on the reading path, because the pipeline should just keep reading the connection. When a timeout occurred, it is hard to distinguish it is really timed out or there is really no response.

The TCP keepalive is a reasonable choice to me. If that also sounds good to you, I would like the merge this PR.

All you considerations are very helpful to me. Thanks again!

Comment (FZambia) [reply pipe.go:460]: Yeah, I see now that you are using small keepalive time, and Go uses 15 secs keepalive by default (not the Linux default which is 7200 secs). I think keepalive will work most of the time, but personally I'd use deadlines as more predictable way to control timeouts. On timeout connection should be closed and all inflight requests should be resolved with an error. Also, I am not sure keepalive helps with timeout which can happen on write to OS network buffer (which is SetWriteDeadline is aim to help with). Though not really sure that's all the case – so I think this pr is fine to merge, any improvements can be done later 👍

Comment (rueian) [reply pipe.go:460]: Oh, thanks. Your concern is right.

I just know that keepalive will not trigger if there is data in the transmission buffer.

So we still need to set WriteDrealine. I will add it later.
